qrenderdoc/Windows/ShaderViewer.cpp

```diff
@@ -1217,16 +1217,63 @@ void ShaderViewer::ConfigureBookmarkMenu()
   QObject::connect(clearAction, &QAction::triggered, [this]() { ClearAllBookmarks(); });
   bookmarkMenu->addAction(clearAction);
 
-  QObject::connect(bookmarkMenu, &QMenu::aboutToShow, [this, nextAction, prevAction, clearAction]() {
-    const bool hasBookmarks = HasBookmarks();
-    nextAction->setEnabled(hasBookmarks);
-    prevAction->setEnabled(hasBookmarks);
-    clearAction->setEnabled(hasBookmarks);
-  });
+  QObject::connect(bookmarkMenu, &QMenu::aboutToShow,
+                   [this, bookmarkMenu, nextAction, prevAction, clearAction]() {
+                     UpdateBookmarkMenu(bookmarkMenu, nextAction, prevAction, clearAction);
+                   });
 
+  bookmarkMenu->addSeparator();
   ui->bookmark->setMenu(bookmarkMenu);
 }
 
+void ShaderViewer::UpdateBookmarkMenu(QMenu *menu, QAction *nextAction, QAction *prevAction,
+                                      QAction *clearAction)
+{
+  // setup permanent actions
+  const bool hasBookmarks = HasBookmarks();
+  nextAction->setEnabled(hasBookmarks);
+  prevAction->setEnabled(hasBookmarks);
+  clearAction->setEnabled(hasBookmarks);
+
+  // remove current bookmark list
+  QList<QAction *> actions = menu->actions();
+  for(auto itAction = actions.rbegin(); itAction != actions.rend() && !(*itAction)->isSeparator();
+      ++itAction)
+    menu->removeAction(*itAction);
+
+  // populate bookmark list
+  ScintillaEdit *cur = currentScintilla();
+  if(!cur)
+    return;
+
+  auto itBookmarks = m_Bookmarks.find(cur);
+  if(itBookmarks == m_Bookmarks.end())
+    return;
+
+  int numAddedBookmarks = 0;
+  for(auto &lineNumber : *itBookmarks)
```

Comment:
Please use the actual type here `sptr_t` rather than `auto`, which I only use for types where the definition is too long-winded like iterators or lambdas.